api/src/unraid-api/plugin/plugin.service.ts

```diff
@@ -1,225 +1,80 @@
-            })
-            .finally(() => {
-                // clear loading state
-                this.loadingPromise = undefined;
-            });
-
-        return this.loadingPromise;
-    }
-
-    private static isPluginFactory(factory: unknown): factory is ConstructablePlugin {
-        return typeof factory === 'function';
-    }
-
-    private static async getPluginFromPackage(pluginPackage: string): Promise<{
-        provider: CustomProvider;
-        pluginInstance: ApiPluginDefinition;
-    }> {
-        const moduleImport = await import(/* @vite-ignore */ pluginPackage);
-        const pluginName = pascalCase(pluginPackage);
-        const PluginFactory = moduleImport.default || moduleImport[pluginName];
-
-        if (!PluginService.isPluginFactory(PluginFactory)) {
-            throw new Error(`Invalid plugin from ${pluginPackage}. Must export a factory function.`);
-        }
-
-        const logger = new Logger(PluginFactory.name);
-        const validation = apiPluginSchema.safeParse(PluginFactory({ store, logger }));
-        if (!validation.success) {
-            throw new Error(`Invalid plugin from ${pluginPackage}: ${validation.error}`);
-        }
-        const pluginInstance = validation.data;
-
-        return {
-            provider: {
-                provide: PluginFactory.name,
-                useValue: pluginInstance,
-            },
-            pluginInstance,
-        };
-    }
+    private static plugins: Promise<ApiNestPluginDefinition[]> | undefined;
 
     static async getPlugins() {
-        /** All api plugins must be npm packages whose name starts with this prefix */
-        const pluginPrefix = 'unraid-api-plugin-';
-        // All api plugins must be installed as dependencies of the unraid-api package
-        /** list of npm packages that are unraid-api plugins */
-        const plugins = getPackageDependencies()?.filter((pkgName) => pkgName.startsWith(pluginPrefix));
-        if (!plugins) {
-            PluginService.logger.warn('Could not load dependencies from the Unraid-API package.json');
-            // Fail silently: Return the module without plugins
-            return [];
-        }
+        PluginService.plugins ??= PluginService.importPlugins();
+        return PluginService.plugins;
+    }
 
-        const failedPlugins: string[] = [];
-        const { data: pluginProviders } = await batchProcess(plugins, async (pluginPackage) => {
+    static async getGraphQlSchemas() {
+        const plugins = (await PluginService.getPlugins()).filter(
+            (plugin) => plugin.graphqlSchemaExtension
+        );
+        const { data: schemas } = await batchProcess(plugins, async (plugin) => {
             try {
-                return await PluginService.getPluginFromPackage(pluginPackage);
+                // safe to assert bc we've filtered the plugins to only include those with graphqlSchemaExtension
+                const schema = await plugin.graphqlSchemaExtension!();
```

Comment:
Should we just add a helper method to filter down the type so we don't need to use assertion here? Not really necessary but would be a bit cleaner than the comment.

.filter(pluginHasSchemaExtension)?